File: src/main/scala/imcode/server/document/index/service/DocumentIndexService.scala
package imcode.server.document.index.service

import java.lang.Long

import com.imcode._
import org.apache.solr.client.solrj.SolrQuery
import org.apache.solr.client.solrj.response.QueryResponse

trait DocumentIndexService {

  def query(solrQuery: SolrQuery): QueryResponse

  /**
   * Updates index asynchronously.
   *
   * @param request index update request.
   */
  def update(request: IndexUpdateOp)

  /**
   * Attempts to run a new index rebuild task if one is not already running.
   *
   * @return current or a new rebuild task.
   */
  def rebuild(): IndexRebuildTask

  /**
   * @return None if index is not empty or Some(attempt to rebuild).
   */
  def rebuildIfEmpty(): Option[IndexRebuildTask] = {
    if (count == 0L) Some(rebuild()) else None
  }

  // todo: optimize
  def count: Long = query(new SolrQuery("*:*")).getResults.getNumFound

  /**
   * @return current rebuild task
   */
  def currentRebuildTaskOpt(): Option[IndexRebuildTask]

  def shutdown()
}


object DocumentIndexService {
  implicit def toManagedResource[R <: DocumentIndexService] = new ManagedResource[R] {
    def close(resource: R) {
      resource.shutdown()
    }

    override def toString = "ManagedResource[_ <: DocumentIndexService]"
  }
}

File: src/main/java/imcode/server/document/index/service/DocumentIndexService.java
package imcode.server.document.index.service;

import org.apache.solr.client.solrj.SolrQuery;
import org.apache.solr.client.solrj.response.QueryResponse;

import java.util.concurrent.Future;

// translated from scala...
public interface DocumentIndexService {

    QueryResponse query(SolrQuery solrQuery);

    /**
     * Updates index asynchronously.
     *
     * @param request index update request.
     */
    void update(IndexUpdateOp request);

    /**
     * Attempts to run a new index rebuild task if one is not already running.
     *
     * @return rebuild future.
     */
    Future rebuild();

    // previously failed indexing not covered here
    default void rebuildIfEmpty() {
        final long foundResults = query(new SolrQuery("*:*")).getResults().getNumFound();
        if (foundResults == 0L) rebuild();
    }

    void shutdown();

    boolean isUpdateDone();

}
